dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/AbstractTrackerDtoValidationHook.java

```diff
@@ -0,0 +1,323 @@
+        Objects.requireNonNull( teAttr, TRACKED_ENTITY_ATTRIBUTE_CANT_BE_NULL );
+
+        String error;
+
+        // We need to do try/catch here since validateValueType() since validateValueType can cast IllegalArgumentException e.g.
+        // on at org.joda.time.format.DateTimeFormatter.parseDateTime(DateTimeFormatter.java:945)
+        try
+        {
+            error = teAttrService.validateValueType( teAttr, attr.getValue() );
+        }
+        catch ( Exception e )
+        {
+            error = e.getMessage();
+        }
+
+        if ( error != null )
+        {
+            ValueType valueType = teAttr.getValueType();
+            errorReporter.addError( newReport( TrackerErrorCode.E1007 )
+                .addArg( valueType.toString() )
+                .addArg( error ) );
+        }
+    }
+
+    protected void validateAttributeUniqueness( ValidationErrorReporter errorReporter,
+        String value,
+        TrackedEntityAttribute trackedEntityAttribute,
+        TrackedEntityInstance trackedEntityInstanceUid,
+        OrganisationUnit organisationUnit )
+    {
+        Objects.requireNonNull( trackedEntityAttribute, TRACKED_ENTITY_ATTRIBUTE_CANT_BE_NULL );
+
+        if ( Boolean.FALSE.equals( trackedEntityAttribute.isUnique() ) )
+            return;
+
+        String error = teAttrService.validateAttributeUniquenessWithinScope(
+            trackedEntityAttribute,
+            value,
+            trackedEntityInstanceUid,
+            organisationUnit );
+
+        if ( error != null )
+        {
+            errorReporter.addError( newReport( TrackerErrorCode.E1064 )
+                .addArg( error ) );
+        }
+    }
+
+    protected void validateGeometry( ValidationErrorReporter errorReporter, Geometry geometry, FeatureType featureType )
+    {
+        Objects.requireNonNull( geometry, GEOMETRY_CANT_BE_NULL );
+
+        if ( featureType == null )
+        {
+            errorReporter.addError( newReport( TrackerErrorCode.E1074 ) );
+            return;
+        }
+
+        FeatureType typeFromName = FeatureType.getTypeFromName( geometry.getGeometryType() );
+
+        if ( FeatureType.NONE == featureType || featureType != typeFromName )
+        {
+            errorReporter.addError( newReport( TrackerErrorCode.E1012 )
+                .addArg( featureType.name() ) );
+        }
+    }
+
+    protected void validateNotes( ValidationErrorReporter reporter, TrackerImportStrategy strategy, List<Note> notes )
+    {
+        for ( Note note : notes )
+        {
+            boolean validUid = CodeGenerator.isValidUid( note.getNote() );
+            if ( !validUid )
+            {
+                reporter.addError( newReport( TrackerErrorCode.E1118 )
+                    .addArg( note.toString() ) );
+            }
+
+            if ( strategy.isCreate() )
+            {
+                //TODO: This looks like a potential performance killer, existence check on every note...
```

Comment:
Not sure how to solve it, anyway it make sense to put it as last in the method

Reply:
If we are checking DB for existance, that implies the data already exists, so we can probably fetch these in the preheater.

I think it's ok to leave it as is for this PR, but we should move this to preheater after.